lib/qos/qos.c

```diff
@@ -0,0 +1,335 @@
+	};
+
+	k_mutex_lock(&ctx_lock, K_FOREVER);
+
+	if (message == NULL) {
+		LOG_WRN("Message cannot be NULL");
+		err = -EINVAL;
+		goto exit;
+	}
+
+	/* Only ACK_REQUIRED messages are added to the internal list. */
+	if (qos_message_has_flag(message, QOS_FLAG_RELIABILITY_ACK_REQUIRED)) {
+		ret = list_append(message);
+		if (ret < 0) {
+			LOG_WRN("No list entries available, error: %d", ret);
+			evt.type = QOS_EVT_MESSAGE_REMOVED_FROM_LIST;
+			notify_event(&evt);
+			err = -ENOMEM;
+			goto exit;
+		}
+
+		ctx.list_internal[ret].message.notified_count++;
+		notify_event(&evt);
+	} else {
+		notify_event(&evt);
+
+		/* If the message does not carry the ACK_REQUIRED flag, its removed from the
+		 * internal list immediately.
+		 */
+		evt.type = QOS_EVT_MESSAGE_REMOVED_FROM_LIST;
+		notify_event(&evt);
+	}
+
+	timer_start();
+
+exit:
+	k_mutex_unlock(&ctx_lock);
+	return err;
+}
+
+int qos_message_remove(uint32_t id)
+{
+	int err;
+
+	k_mutex_lock(&ctx_lock, K_FOREVER);
+
+	err = list_remove(id);
+	if (err) {
+		goto exit;
+	}
+
+	/* If the removed message is the last in the pending list, we stop the internal timer
+	 * and clear the retry counter.
+	 */
+	if (sys_slist_is_empty(&ctx.pending_list)) {
+		LOG_WRN("QoS list is empty!");
+		ctx.retry_count = 0;
+		k_work_cancel_delayable(&ctx.timeout_handler_work);
+	}
+
+exit:
+	k_mutex_unlock(&ctx_lock);
+	return err;
+}
+
+void qos_message_print(const struct qos_data *message)
+{
+	LOG_DBG("Notified count: %d", message->notified_count);
+	LOG_DBG("Message heap_allocated: %d", message->heap_allocated);
+	LOG_DBG("Message ID: %d", message->id);
+	LOG_DBG("Message Buffer pointer: %p", message->data.buf);
+	LOG_DBG("Message Buffer length: %d", message->data.len);
+	LOG_DBG("Message Flags: %x", message->flags);
+	LOG_DBG("Message type: %d", message->type);
+}
+
+bool qos_message_has_flag(const struct qos_data *message, uint32_t flag)
+{
+	return ((message->flags & flag) == flag) ? true : false;
+}
+
+uint16_t qos_message_id_get_next(void)
+{
+	if (message_id_next > UINT16_MAX) {
+		message_id_next = QOS_MESSAGE_ID_BASE;
+	}
+
+	message_id_next++;
+
+	return message_id_next;
+}
+
+void qos_message_notify_all(void)
+{
+	struct qos_metadata *node = NULL, *next_node = NULL;
+	struct qos_evt evt = {
+		.type = QOS_EVT_MESSAGE_TIMER_EXPIRED,
+	};
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&ctx.pending_list, node, next_node, header) {
+		node->message.notified_count++;
+		evt.message = node->message;
+		notify_event(&evt);
+	};
+}
+
+void qos_message_remove_all(void)
+{
+	struct qos_metadata *node = NULL, *next_node = NULL, *node_prev = NULL;
+	struct qos_evt evt = {
+		.type = QOS_EVT_MESSAGE_REMOVED_FROM_LIST,
+	};
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&ctx.pending_list, node, next_node, header) {
+		evt.message = node->message;
+		notify_event(&evt);
+		memset(&node->message, 0, sizeof(struct qos_data));
+		sys_slist_remove(&ctx.pending_list, &node_prev->header, &node->header);
+		node_prev = node;
```

Comment:
Is this safe? We've just removed the `node` from the list, I don't think it's ok to use it as a previous node reference. Maybe it'd safer to use something like `while (node = sys_slist_peek_head(&ctx.pending_list)` and keep removing head element (prev set to NULL)?

Reply:
Good point. Ill do some testing and see if I can come up with an alternative using your suggestion.

Reply:
I replaced the `sys_slist_peek_head` call with the `sys_slist_find_and_remove` flavour. It doesn't require the node prev value.